packages/compartment-mapper/src/import-hook.js

```diff
@@ -398,3 +405,256 @@ export const makeImportHookMaker = (
+ *
+ * @param {SyncReadPowers} readPowers
+ * @param {string} baseLocation
+ * @param {MakeImportNowHookMakerOptions} options
+ * @returns {ImportNowHookMaker|undefined}
+ */
+export function makeImportNowHookMaker(
+  readPowers,
+  baseLocation,
+  {
+    sources = Object.create(null),
+    compartmentDescriptors = Object.create(null),
+    computeSha512 = undefined,
+    searchSuffixes = nodejsConventionSearchSuffixes,
+    sourceMapHook = undefined,
+    exitModuleImportNowHook,
+  },
+) {
+  // Set of specifiers for modules (scoped to compartment) whose parser is not
+  // using heuristics to determine imports.
+  /** @type {Map<string, Set<string>>} compartment name ->* module specifier */
+  const strictlyRequired = new Map();
+
+  /**
+   * @param {string} compartmentName
+   */
+  const strictlyRequiredForCompartment = compartmentName => {
+    let compartmentStrictlyRequired = strictlyRequired.get(compartmentName);
+    if (compartmentStrictlyRequired !== undefined) {
+      return compartmentStrictlyRequired;
+    }
+    compartmentStrictlyRequired = new Set();
+    strictlyRequired.set(compartmentName, compartmentStrictlyRequired);
+    return compartmentStrictlyRequired;
+  };
+
+  /**
+   * @type {ImportNowHookMaker}
+   */
+  const makeImportNowHook = ({
+    packageLocation,
+    packageName: _packageName,
+    entryCompartmentDescriptor,
+    parse,
+    compartments,
+  }) => {
+    const compartmentDescriptor = compartmentDescriptors[packageLocation] || {};
+
+    // this is necessary, because there's nothing that prevents someone from calling this function with the entry compartment.
+    assert(
+      compartmentDescriptor !== entryCompartmentDescriptor,
+      'Cannot create an ImportNowHook for the entry compartment',
+    );
+
+    // this is not strictly necessary, since the types should handle it.
+    assert(
+      parse[Symbol.toStringTag] !== 'AsyncFunction',
+      'async parsers are unsupported',
+    );
+
+    packageLocation = resolveLocation(packageLocation, baseLocation);
+    const packageSources = sources[packageLocation] || Object.create(null);
+    sources[packageLocation] = packageSources;
+    const { modules: moduleDescriptors = Object.create(null) } =
+      compartmentDescriptor;
+    compartmentDescriptor.modules = moduleDescriptors;
+
+    let { policy } = compartmentDescriptor;
+    policy = policy || Object.create(null);
+
+    // associates modules with compartment descriptors based on policy
+    // which wouldn't otherwise be there
+    if ('packages' in policy && typeof policy.packages === 'object') {
+      for (const [pkgName, policyItem] of entries(policy.packages)) {
+        if (
+          !(pkgName in compartmentDescriptor.modules) &&
+          pkgName in compartmentDescriptor.scopes &&
+          policyItem
+        ) {
+          compartmentDescriptor.modules[pkgName] =
+            compartmentDescriptor.scopes[pkgName];
+        }
+      }
+    }
```

Comment:
I am unsure if this is correct, but it creates links between compartment descriptors based only on policy--where Endo would not have detected them otherwise.

Reply:
less code and better compatibility. Sounds good.
Also, explicit opt-in would allow an early failure if ReadPowers or transforms don't match the intent to enable dynamic require.